### PR TITLE
[ST-Engine] [Merge] Merge SeaTunnel Engine Branch to dev

### PR DESCRIPTION
**Pull request by Hisoka-X**
## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -198,6 +201,8 @@
         <snappy-java.version>1.1.8.3</snappy-java.version>
         <checker.qual.version>3.10.0</checker.qual.version>
         <awaitility.version>4.2.0</awaitility.version>
+        <!--  SeaTunnel Engine use     -->
+        <hazelcast.version>5.1</hazelcast.version>
     </properties>
```
Comment (CalvinKirs) [inline pom.xml]: The package management of the engine should be handed over to the engine alone, it does not belong to the dependencies required by the public module
The package management of the engine should be handed over to the engine alone, it does not belong to the dependencies required by the public module

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/RetryUtils.java [hunk 2]
```diff
@@ -34,21 +34,25 @@ public static <T> T retryWithException(Execution<T, Exception> execution, RetryM
         if (retryMaterial.getRetryTimes() < 0) {
             throw new IllegalArgumentException("Retry times must be greater than 0");
         }
+        Exception lastE;
```
Comment (CalvinKirs) [inline seatunnel-common/src/main/java/org/apache/seatunnel/common/utils/RetryUtils.java]: Rename this variable

**Review by CalvinKirs — DISMISSED** [review]
And the project is full of various RuntimeException to be resolved. If there is no big problem, it is best to merge this PR as soon as possible

**Comment by CalvinKirs** [comment]
Some packaging and the like need to be tested later and assigned to me @CalvinKirs

**Review by CalvinKirs — DISMISSED** [review]

**Comment by CalvinKirs** [comment]
> Some packaging and the like need to be tested later and assigned to me @CalvinKirs

And some CI need to be refactor

**Comment by CalvinKirs** [comment]
In addition, we also need to solve some bugs of the `sonar`, we can create a separate PR to solve them after merging.

File: seatunnel-server/seatunnel-server-common/src/main/java/org/apache/seatunnel/server/common/SeatunnelErrorEnum.java [hunk 3]
```diff
@@ -43,7 +43,7 @@ public enum SeatunnelErrorEnum {
     UNSUPPORTED_OPERATION(99996, "unsupported operation", "This operation [%s] is not supported now."),
     HTTP_REQUEST_FAILED(99997, "http request failed", "Http request failed, url is %s"),
     ILLEGAL_STATE(99998, "illegal state", "%s"),
-    UNKNOWN(99999, "unknown exception", "Unknown exception")
+    UNKNOWN(99999, "unknown exception", "%s")
```
Comment (ashulin) [inline seatunnel-server/seatunnel-server-common/src/main/java/org/apache/seatunnel/server/common/SeatunnelErrorEnum.java:46]: Changes that are not part of the engine

**Review by ashulin — DISMISSED** [review]

**Review by CalvinKirs — APPROVED** [review]

**Review by hailin0 — APPROVED** [review]

**Review by EricJoy2048 — APPROVED** [review]
+1

**Pull request merged**